pkg/netns/netns_linux.go

```diff
@@ -0,0 +1,200 @@
+func DoByPinPath(netNSPath string, runInside func(*NetNSHandle) error) error {
+	fd, err := unix.Open(netNSPath, unix.O_RDONLY|unix.O_CLOEXEC, 0)
+	if err != nil {
+		return err
+	}
+	// We aren't keeping this and we don't own the pin or the netns it refers to
+	// so defer closure.
+	defer func() {
+		err = errors.Join(err, unix.Close(fd))
+	}()
+
+	localHandle := NetNSHandle{
+		fd: int(fd),
+	}
+	err = localHandle.Do(runInside)
+	return err
+}
+
+// FD returns the underlying file descriptor of the NetNSHandle
+// Note that FDs should not be used for comparing or testing equality, for that
+// see `equals`
+func (h *NetNSHandle) FD() int {
+	return h.fd
+}
+
+// Close closes the handle to the network namespace. Note that this does not guarantee
+// the network namespace itself will be destroyed or closed, which will happen when all references to it
+// (including the one being closed here) are gone.
+func (h *NetNSHandle) Close() error {
+	if err := unix.Close(h.fd); err != nil {
+		return err
+	}
+	h.fd = -1
+	return nil
+}
+
+// Do runs the provided func within a new thread+netns, without changing the current thread's netns
+//
+// Note that since network namespaces are set *per thread*, the thread the closure is running inside
+// will be locked with `runtime.LockOSThread()` during this operation.
+//
+// This means the provided closure should _not_ call runtime.UnlockOSThread() at all,
+// or else undefined behavior should be expected.
+func (h *NetNSHandle) Do(runInside func(*NetNSHandle) error) error {
+	var wg sync.WaitGroup
+	wg.Add(1)
+
+	// Start the callback in a new green thread so that if we later fail
+	// to switch the namespace back to the original one, we can safely
+	// leave the thread locked to die without a risk of the current thread
+	// left lingering with incorrect namespace.
+	var innerError error
+	go func() {
+		// save a handle to current network namespace
+		// and lock the thread - this will force this goroutine to be
+		// scheduled in a different thread, so we should be safe to mutate the netns
+		restore, err := saveRestoreCurrentThreadNetns()
+		if err != nil {
+			innerError = err
+			return
+		}
+		defer restore()
+		defer wg.Done()
+		unix.Setns(h.fd, unix.CLONE_NEWNET)
+		innerError = runInside(h)
+	}()
+	wg.Wait()
+
+	return innerError
+}
+
+// NOTE If we ever need to test equality of netNSes outside of test code,
+// make this public.
+func (h *NetNSHandle) equal(other *NetNSHandle) bool {
+	if h.FD() == other.FD() {
+		return true
+	}
+
+	// Just because the FDs are not equal doesn't mean they don't both refer
+	// to the same underlying netns, so check inode too.
+	var s1, s2 unix.Stat_t
+	if err := unix.Fstat(int(h.FD()), &s1); err != nil {
+		return false
+	}
+	if err := unix.Fstat(int(other.FD()), &s2); err != nil {
+		return false
+	}
+	return (s1.Dev == s2.Dev) && (s1.Ino == s2.Ino)
+}
+
+// saveRestoreCurrentThreadNetns returns a cleanup func (which callers should `defer`)
```

Comment:
This only documents what the cleanup function does and is redundant with the second paragraph. I'd call this reset/revert instead of cleanup, which sort of implies removing files or other resources.